### PR TITLE
Yum versionlock support

### PR DESCRIPTION
**Pull request by riemers**
##### SUMMARY
This module adds support to lock packages in their version to prevent these from updating by yum.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
yum_versionlock

##### ADDITIONAL INFORMATION
Please see: https://github.com/ansible/ansible/pull/41778 for more information, this is just a drag and drop to get the ball rolling.

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @felixfontein @Andersson007 @gundalow Could you please review this? Thanks in advance.

**Comment by Andersson007** [comment]
can't this functionality be added to the `yum` module?

**Comment by riemers** [comment]
If that is possible, sure. I just found something i needed which was stale for 2 years then i moved it over as per request.  I don't normally play around in this area, so if someone could help out or give some pointer so it can become a joined effort perhaps. I know we are all busy, so am i. Trying to make the best of it.

**Comment by mator** [comment]
could be probably added to yum, something like `state: lock` and `state: unlock`

**Comment by Andersson007** [comment]
for example,
```
yum:
   pkg: blahbla
   state: locked
```
or something
@felixfontein , your opinion?

**Comment by Andersson007** [comment]
@mator , yep, you were a bit faster:)

**Comment by Akasurde** [comment]
I disagree. Since `yum` module is overloaded with features. I think new module will keep this feature simple and sweet.

**Comment by florianpaulhoberg** [comment]
I think yum base module is also too complex, now. I think it may be more overseen / not found within that module. On the other hand it's related to this module and just a state 'locked' or 'unlocked' (instead of installed, etc.) would be even nice.

**Comment by Akasurde** [comment]
I agree this is related to `yum` but keeping a separate module will make module maintainable and future proof.

**Comment by Andersson007** [comment]
For me personally, it doesn't look bad to add `locked/unlocked` states to the yum module: a couple of rows in the documentation and two more examples:)
Despite the module is pretty complex now, it looks logical to have this option there.

But it can be a separate module as well.
Anyway, the module is not ok in this condition because:
1. we need to have it idempotent. It doesn't support the check mode also
AFAIK, there is `yum versionlock list` command, it'll help to solve the problem
2. there aren't tests

**Comment by riemers** [comment]
If it is suppose to go into the yum module would add a requirement of having the yum-plugin-versionlock package installed no?

**Comment by Andersson007** [comment]
@riemers probably yes.
We need more votes about yum-inclusion / a separate module (the last one is ok imo, but it needs to have check mode support and be idempotent in general - because it looks like it can be)

**Comment by Andersson007** [comment]
i'll ask guys in irc

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: yum_versionlock
+version_added: 2.7
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:19]:
```suggestion
version_added: 2.10
```

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+short_description: Locks/Unlocks an installed package(s) from being updates by (yum) package manager.
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:20]:
```suggestion
short_description: Locks/Unlocks an installed package(s) from being updates by yum package manager
```

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+options:
+  package:
+    description:
+      - Package name(s) like C(httpd). Multiple packages are supported by adding them by whitespaces.
+  state:
+    description:
+      - Whether to lock C(present) or unlock C(absent) package(s).
+    choices: [ present, absent ]
+    default: present
+# informational: requirements for nodes
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:32]:
```suggestion
```

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+requirements:
+- yum
+- yum-versionlock
+author:
+    - '"Florian Paul Hoberg (@florianpaulhoberg)" <florian.hoberg@credativ.de>'
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:37]:
```suggestion
    - Florian Paul Hoberg (@florianpaulhoberg)
```

**Comment by felixfontein** [comment]
CI tests are actually requried for new modules :) Probably integration tests would be best.

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+'''
+
+EXAMPLES = '''
+- name: Prevent Apache / httpd from being updated
+  yum_versionlock:
+    state: present
+    package: httpd
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:44]:
```suggestion
    package: httpd

```

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+- name: Prevent multiple packages from being updated
+  yum_versionlock:
+    state: present
+    package: httpd nginx haproxy curl
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:48]:
```suggestion
    pkg:
    - httpd
    - nginx
    - haproxy
    - curl

```

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+DOCUMENTATION = '''
+---
+module: yum_versionlock
+version_added: 2.7
+short_description: Locks/Unlocks an installed package(s) from being updates by (yum) package manager.
+description:
+     - This module adds installed packages to yum versionlock to prevent it from being updated.
+options:
+  package:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:24]:
```suggestion
  pkg:
```
`yum` module uses `pkg` alias, imo, would be cool to have the same this as well

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    description:
+      - Package name(s) like C(httpd). Multiple packages are supported by adding them by whitespaces.
+  state:
+    description:
+      - Whether to lock C(present) or unlock C(absent) package(s).
+    choices: [ present, absent ]
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:30]:
```suggestion
    choices: [ present, absent ]
    type: str
```

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+---
+module: yum_versionlock
+version_added: 2.7
+short_description: Locks/Unlocks an installed package(s) from being updates by (yum) package manager.
+description:
+     - This module adds installed packages to yum versionlock to prevent it from being updated.
+options:
+  package:
+    description:
+      - Package name(s) like C(httpd). Multiple packages are supported by adding them by whitespaces.
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:26]:
```suggestion
      - Package name or a list of packages.
      type: list
```
imo, more logical to have a list here.
So we could pass values separated by commas or using the yml list syntax (see the changed example).

**Comment by Andersson007** [comment]
I haven't looked at the code yet (will see after the mentioned stuff is implemented:)
any questions? feel free to ask!

**Comment by riemers** [comment]
Ok i am willing to give it a shot, but i don't often work with python. That said, do you have a similar module which has these steps or a past PR so i can take a peek. That usually gets me somewhere, i started the move, should at least do my best.

**Comment by Akasurde** [comment]
@riemers Could you please merge https://github.com/riemers/community.general/pull/1? I included all required changes from my side.

**Comment by riemers** [comment]
@Akasurde done, although another error, i though you only need a rebase when according to the url files are changed which are in git or too far behind. Since it is a new file, what triggered it? Can i resolve this from within github too? Not at home.

**Comment by Akasurde** [comment]
I rebased patch-1 and force pushed to your patch-1 branch. Now we are good to go. Could you please add integration test for this change? Thanks.

**Comment by riemers** [comment]
Looking in to it, might take me some time but i'am on it.

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:15]:
```suggestion
```
No longer needed, please, remove this

File: plugins/modules/packaging/os/yum_versionlock.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: yum_versionlock
+version_added: 2.10
```
Comment (Andersson007) [inline plugins/modules/packaging/os/yum_versionlock.py:13]:
```suggestion
```

**Comment by Andersson007** [comment]
any updates on this?

**Review by Andersson007 — COMMENTED** [review]
should be `version_added: 1.3.0` for now

**Comment by riemers** [comment]
To be frank, I don't have the time/effort it has to finish it. Can just as well be closed then unless someone else wants it.

**Comment by Andersson007** [comment]
@riemers , can anyone finish the pr adding oneself as a coauthor?

**Comment by riemers** [comment]
If "anybody" wants to, they can go ahead. I can click around if needed but that is about it for now for me.

File: plugins/modules/packaging/os/yum_versionlock.py (hunk 11 above)
Comment (aminvakil) [inline plugins/modules/packaging/os/yum_versionlock.py:13]:
```suggestion
version_added: 2.0.0
```

**Review by aminvakil — COMMENTED** [review]
I want to try adding an integration test for this module and make any other changes this needs, should I open a PR to https://github.com/riemers/community.general/tree/patch-1 ?

**Comment by felixfontein** [comment]
@aminvakil since @riemers doesn't want to work on this anymore, it's probably easier to create a new branch based on their branch and work on that one (i.e. add new commits).

**Comment by riemers** [comment]
It is not that i don't want to maintain it, i can handle that. But i cannot code/test on it anymore. I try to do the best i can.

**Comment by dmsimard** [comment]
> It is not that i don't want to maintain it, i can handle that. But i cannot code/test on it anymore. I try to do the best i can.

Thanks for your effort @riemers, it's appreciated !

@aminvakil took over and submitted a new PR for this: https://github.com/ansible-collections/community.general/pull/1405

I will close this one.

**Pull request closed without merging**